### ISSUE TITLE
Google Meet: "Your browser can't share your screen"

### ISSUE THREAD

**Issue by renatolond** [Eloston/ungoogled-chromium#601]
**Describe the bug**
When using Google Meet, if I try to share my screen I get an error from google meet saying "Your browser can't share your screen".

**To Reproduce**
1. Go to 'Google Meet'
2. Start a new meeting
3. Click on "Share my screen", "Share entire screen"

**Expected behavior**
I expected my screen to be shared, as it is in Chromium.

**Environment (please complete the following information):**
 - OS and version: ArchLinux
 - ungoogled-chromium version: 70.0.3538.77 (from AUR, ungoogled-chromium-bin)

**Additional context**
It seems brave used to have an issue sharing the screen with hangouts, here: https://github.com/brave/browser-laptop/issues/3392
I don't know if they can be related.

**Comment by ian-moone** [comment]
If it depends on _hangouts_, then I guess that's the expected behavior, as `enable_hangout_services_extension` [is disabled](https://github.com/Eloston/ungoogled-chromium/blob/master/config_bundles/common/gn_flags.map#L4).

**Comment by Eloston** [comment]
Based on [this article](https://zapier.com/blog/google-hangouts-meet-guide/), it looks like Google Meet is just redesigned Hangouts. In that case, what @ian-moone should be correct. I don't want to change the GN flag or add extra code (e.g. patches for a flag or something) since it clearly goes against the spirit of this project. I will at most allow instructions in the Wiki to get this working (if anyone is inclined to write them in the first place).

If this is not due to the Hangouts extension, please let us know.

**Comment by renatolond** [comment]
Thanks for the response!
I was not sure if there was any google component involved, but it does make sense that they would be using their own stuff.
I will check what would be needed and if I find it I will document it on the wiki :)

@Atavic having to use google meet for work does not make me want to give google control over every other website I access, hence the issue.

**Comment by Atavic** [comment]
Ok, now understand your position. Corporations like MS, Apple, Google and Amazon are improving the grip on their customers every minute. My only idea (for an issue that I met myself) is to install more browsers or use more profiles for specific needs. Even Virtual Machines can be useful.

**Comment by csalvato** [comment]
Why can't we, as users, decide whether or not to enable this extension? Why is this marked as wontfix?

**Comment by Atavic** [comment]
Because it is disabled with this [option](https://github.com/Eloston/ungoogled-chromium/blob/master/config_bundles/common/gn_flags.map#L4) when chromium is built from source.

**Comment by Eloston** [comment]
@csalvato I don't understand how supporting this "extension" (it's actually code built-in to the browser) will contribute to the goals of this project. Also, if you need the extension, changing the flag that @Atavic mentioned and re-building the browser will fix the issue. Could you explain your use-case in more depth? Why would you need a Google-specific component in a browser designed to remove Google integration?

**Comment by csalvato** [comment]
@Eloston I am not super intimate with the goals of the project, but I would image that one of the goal is to actually have the browser get adopted.  If that's the case, then this extension being excluded creates a massive barrier for many people who rely on screen sharing with Google Meet.

Google Meet is one of the most popular and pervasive video meeting and screen sharing tools for remote teams. Every team I've been on since 2010 has used Google Hangouts or Google Meet daily.

If I need to load up Chrome to run my work meetings, then switch to Brave for everything else, that's a level of complexity that is too much for me, personally. And if the goal is mass adoption, not supporting Google Meet screensharing will be in the way for many more people who don't want to deal with a wonky, persnickety browser.

I can use Google Meet screen sharing on Safari, Chrome and Firefox.  But I can't use it on Brave. If one of the goals of the project is to not support an integral remote-working use case, I guess Brave isn't for me. 😢

**Comment by csalvato** [comment]
**Re: rebuilding the browser:** I wouldn't even know how to rebuild the browser from source just to get Google Meet working.  If that's the solution to the problem, I think you'll find a barrier to getting more people to use Brave.

**Comment by csalvato** [comment]
@antonin-lebrard Oh wow! My mistake. I was linked here from another issue on the Brave project.  How embarrassing :)

**Comment by csalvato** [comment]
In that context, that makes total sense.

**Comment by Eloston** [comment]
@csalvato First, I would like to note that we are *not* affiliated with Brave. It just so happens that both projects share similar intentions and goals.

Anyway, while I would like to have this project grow, it is secondary to the main objective of this project: to remove integration with Google services. Based on that fact alone, supporting a feature which *only* benefits the usage of Google services is contradictory. However, I understand that there are a variety of valid use cases that does not exclude Google completely, so I generally don't outright reject suggestions like these. There is no clear line to draw here.

After investigating the code for the Hangout service extension, I will retract my earlier statement about rejecting this completely. I think it is possible to have a flag that can enable this extension to be loaded, but not bundle the code in the browser. However, I do not have any interest in working on this, nor do I suspect many others here will given my experience thus far.

**Comment by DashDashDashDashDash** [comment]
Is there any simpler way of allowing this than changing the flag in `flags.gn` and rebuilding the browser? I believe having the flexibility to change this extension is something handy to have. Surely not something important, but in my case, using ungoogled chromium with the hangouts extension is significantly better than having to use regular google chrome with all of its problems.
I do like the idea of some documentation on it, since clarity on this issue is a big obstacle, as I've seen.

**Comment by dndrsn** [comment]
I just tried a build with `enable_hangout_services_extension=true` set in flags.gn and unfortunately it didn't work for me to enable screen sharing in Hangouts.  The behavior is unchanged.  Has anyone been able to get this working?

**Comment by PF4Public** [comment]
@dndrsn see the mentioned issue (PF4Public/gentoo-overlay#42), no solution this far.

**Comment by resoli** [comment]
> I just tried a build with `enable_hangout_services_extension=true` set in flags.gn and unfortunately it didn't work for me to enable screen sharing in Hangouts. The behavior is unchanged. Has anyone been able to get this working?

Same here. We would like to adopt ungoogled in our organization, but the inability to do screen sharing in G. Meet is a showstopper. If someone has some advice, it would be nice. Rebuilding from source is not a problem.

**Comment by Altitude-Dashboard** [comment]
I also need Google for video sharing, it's work related.

**Comment by PF4Public** [comment]
> It doesn't work in Chromium 88.0.4324.182

With which flags do you compile it and "Meet" does not work? Like I've mentioned in the related bug, I've been told, that hangouts do function (at least if built for Gentoo from my overlay).

**Comment by Altitude-Dashboard** [comment]
By "Meet" I was talking about meet.google.com as this feature is now free and everybody is using it instead of the old Hangouts. I am not able the share the screen but conferencing works.
I have not compiled it myself, I used the binaries from chromium.woolyss.com for MacOS. I will try building it myself even if I am not an expert but I did this kind of thing before. I am a bit confused on which flags do I need to switch on to make it work.

**Comment by PF4Public** [comment]
> which flags do I need to switch on to make it work

You should probably start with `enable_hangout_services_extension=true`

**Comment by aspieln3r** [comment]
Google meet works in firefox so it should be possible to bring this to ungoogled chromium also somehow without using google code?

**Comment by Altitude-Dashboard** [comment]
@aspieln3r you need to recompile it yourself with enable_hangout_services_extension=true
I have not tested with Eloston version, I have tested with the official Chromium (the official Chromium is also not officially supporting Hangouts/Meet) and this is solving the problem. Maybe somebody needs to create a fork because the goal of this project is to "ungoogle" everything and somehow this contradicts its purpose.

**Comment by resoli** [comment]
> @aspieln3r you need to recompile it yourself with enable_hangout_services_extension=true

I have built uc with that flag, and doesn't make any difference.

**Comment by resoli** [comment]
>  Maybe somebody needs to create a fork because the goal of this project is to "ungoogle" everything and somehow this contradicts its purpose.
I agree, (unfortunately, uc is wonderful) we are out of scope here.

**Comment by PF4Public** [comment]
> doesn't make any difference

You may try other flags, for example `enable_remoting` and/or `enable_media_remoting`

**Comment by resoli** [comment]
> You may try other flags, for example `enable_remoting` and/or `enable_media_remoting`

Will do and report. Thanks!

**Comment by Altitude-Dashboard** [comment]
For the official Chromium I used:

vi out/Default/args.gn
is_debug = false
symbol_level = 0
ffmpeg_branding = "Chrome"
proprietary_codecs = true
enable_hangout_services_extension = true

Maybe this helps.

**Comment by resoli** [comment]
@Altitude-Dashboard many thanks!!!

**Comment by resoli** [comment]
> For the official Chromium I used:
> 
> vi out/Default/args.gn
> is_debug = false
> symbol_level = 0
> ffmpeg_branding = "Chrome"
> proprietary_codecs = true
> enable_hangout_services_extension = true
> 
> Maybe this helps.

I'm trying latest `ubuntu-focal` flavour of `ungoogled-chromium-debian`:
```
ninja -j4 -C out/Release chrome chrome_sandbox content_shell chromedriver
ninja: Entering directory `out/Release'
[9478/44037] CXX obj/base/i18n/icu_util.o
FAILED: obj/base/i18n/icu_util.o
clang++-11 -MMD -MF obj/base/i18n/icu_util.o.d -DBASE_I18N_IMPLEMENTATION -DUSE_UDEV -DUSE_AURA=1 -DUSE_GLIB=1 -DUSE_NSS_CERTS=1 -DUSE_X11=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -DNO_UNWIND_TABLES -D_GNU_SOURCE -DCR_CLANG_REVISION=\"llvmorg-12-init-11462-g418f18c6-1\" -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D_FORTIFY_SOURCE=2 -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_40 -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_40 -DUSING_SYSTEM_ICU=1 -DICU_UTIL_DATA_IMPL=ICU_UTIL_DATA_STATIC -DUCHAR_TYPE=uint16_t -DU_IMPORT=U_EXPORT -I../.. -Igen -Igen/shim_headers/icui18n_shim -Igen/shim_headers/icuuc_shim -I../../third_party/perfetto/include -Igen/third_party/perfetto/build_config -Igen/third_party/perfetto -Igen/shim_headers/zlib_shim -Igen/shim_headers/libevent_shim -I../../third_party/ced/src -I../../third_party/abseil-cpp -I../../third_party/boringssl/src/include -I../../third_party/protobuf/src -Igen/protoc_out -fno-strict-aliasing --param=ssp-buffer-size=4 -fstack-protector -fno-unwind-tables -fno-asynchronous-unwind-tables -fPIC -pthread -fcolor-diagnostics -fmerge-all-constants -fcrash-diagnostics-dir=../../tools/clang/crashreports -mllvm -instcombine-lower-dbg-declare=0 -m64 -march=x86-64 -Wno-builtin-macro-redefined -D__DATE__= -D__TIME__= -D__TIMESTAMP__= -Xclang -fdebug-compilation-dir -Xclang . -no-canonical-prefixes -Wall -Wextra -Wimplicit-fallthrough -Wunreachable-code -Wthread-safety -Wextra-semi -Wno-missing-field-initializers -Wno-unused-parameter -Wno-c++11-narrowing -Wno-unneeded-internal-declaration -Wno-undefined-var-template -Wno-psabi -Wno-ignored-pragma-optimize -Wno-implicit-int-float-conversion -Wno-final-dtor-non-final-class -Wno-builtin-assume-aligned-alignment -Wno-deprecated-copy -Wno-non-c-typedef-for-linkage -Wmax-tokens -fno-omit-frame-pointer -g0 -ftrivial-auto-var-init=pattern -fvisibility=hidden -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare -Wexit-time-destructors -fno-ident -fdata-sections -ffunction-sections -Wno-shorten-64-to-32 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -std=c++14 -fno-trigraphs -Wno-trigraphs -fno-exceptions -fno-rtti -fvisibility-inlines-hidden -resource-dir=/usr/lib/llvm-11/lib/clang/11.0.0 -resource-dir=/usr/lib/llvm-11/lib/clang/11.0.0 -O2 -fdebug-prefix-map=/home/resolir/git/ungoogled-chromium/build/src=. -fstack-protector-strong -Wformat -Werror=format-security -Wno-conversion -Wno-unused-function -Wno-unused-variable -Wno-unused-private-field -Wno-deprecated-declarations -Wno-unreachable-code -Wno-unknown-pragmas -fno-delete-null-pointer-checks -c ../../base/i18n/icu_util.cc -o obj/base/i18n/icu_util.o
../../base/i18n/icu_util.cc:379:10: error: use of undeclared identifier 'UTRACE_UBRK_CREATE_CHARACTER'
    case UTRACE_UBRK_CREATE_CHARACTER:
         ^
../../base/i18n/icu_util.cc:383:10: error: use of undeclared identifier 'UTRACE_UBRK_CREATE_SENTENCE'
    case UTRACE_UBRK_CREATE_SENTENCE:
         ^
../../base/i18n/icu_util.cc:387:10: error: use of undeclared identifier 'UTRACE_UBRK_CREATE_TITLE'
    case UTRACE_UBRK_CREATE_TITLE:
         ^
../../base/i18n/icu_util.cc:391:10: error: use of undeclared identifier 'UTRACE_UBRK_CREATE_WORD'
    case UTRACE_UBRK_CREATE_WORD:
         ^
../../base/i18n/icu_util.cc:425:10: error: use of undeclared identifier 'UTRACE_UBRK_CREATE_LINE'
    case UTRACE_UBRK_CREATE_LINE: {
         ^
../../base/i18n/icu_util.cc:451:10: error: use of undeclared identifier 'UTRACE_UBRK_CREATE_BREAK_ENGINE'
    case UTRACE_UBRK_CREATE_BREAK_ENGINE: {
         ^
6 errors generated.
```

**Comment by deadlysyn** [comment]
love uc and trying hard to figure out how to keep it around since "the day job" requires meet/hangouts (at least it's not webex or zoom).

since this project has a lot of other things disabled out of the box that can be enabled with a flag, i don't think having one more setting uc users need available but off by default violates project goals.

came here hoping i "just" need to recompile (that's more fun in winter as an extra space heater since it pegs all more cores for quite awhile ;-) ) but it looks like that doesn't even work. don't want chrome around just for this, may have to install vivaldi just for hangouts while hoping for a better solution. (vivaldi is great in many ways, i'm just ocd enough multiple browsers bother me.)

**Comment by noneistaken** [comment]
Currently getting this https://pastebin.com/E6Yatia0 while trying to build `ungoogled-chromium-debian`. Anyone has a clue on how to fix this?

**Comment by pdfrod** [comment]
I would like to keep using ungoogled-chromium but I still need some way of using Google Meet, because that's what is used at work. I'm forced to go back to Google Chrome, which is also contrary to the spirit of this project in my opinion.

Would you be willing to accept a pull request giving users the option to easily enable Google Meet?

**Comment by PF4Public** [comment]
> Would you be willing to accept a pull request giving users the option to easily enable Google Meet?

Have you found a definite solution to this issue?

**Comment by pdfrod** [comment]
No. I wanted to get the mainteners' blessing before spending a lot of of time working on a pull request that might get rejected.

**Comment by PF4Public** [comment]
> before spending a lot of of time working on a pull request

Your best bet would be to first identify the issue itself, then consider what kind of changes are needed to fix it and only then someone could asses your proposal w.r.t. mergeability into ungoogled-chromium or contrib.

**Comment by pdfrod** [comment]
Yeah, but the issue was closed with the justification that fixing it would be against the goals of the project, so I'm not even sure if there's interest in fixing it.

**Comment by PF4Public** [comment]
That might've been a disappointing resolution, but please consider the following:

This feature obviously falls out of the scope of the project and might even spoil all the efforts thus far attained. This feature is thus very unlikely to be used by any maintainer of ungoogled-chromium, which in turn suggests that it is unlikely that any maintainer of ungoogled-chromium would even consider investigating the possibilities. I mean I wouldn't be willing to debug something I do not use, right? That being said, I can find a number of very similar (in this regard) issues on the bug-tracker which are left in mid-air due to the same reasons.

The reason for this suspended state is that no one took the lead and investigated the real cause of why some features cease functioning after ungoogled-chromium patches are applied. If this would have been done, it would be easy to decide which category this fix belongs to: a) conditional branching in runtime without any other changes; b) build time fixes which involve non-default gn flags, selective patch application / domain substituitions etc. Run-time flag would be  probably a merge-able solution, but a build time solution should be applied manually upon build time. We cannot possibly build a ton of varieties for obvious reasons, which means those build-time solutions are worthy to be included in wiki or in contrib to empower users with similar requirements to build their variation for their specific use-case. Why would anyone be against that? So if one is after a solution, there is always one. 

In other words: if one aims to force some unrelated (contradictory even) feature into the repository so that it would be further propagated into the resulting binaries, sorry, no way. On the other hand if one is up for a solution, which one would use and share this solution for others in need — that's a big yes!

**Comment by pdfrod** [comment]
Thanks for the detailed explanation. I'll do some digging and report back when I have some news.

**Comment by whitequark** [comment]
You can make screen sharing work in Google Meet on any Chromium if you go to DevTools and change your User-Agent string to Microsoft Edge (Chromium) in the drop-down box.

**Comment by jaimet** [comment]
> You can make screen sharing work in Google Meet on any Chromium if you go to DevTools and change your User-Agent string to Microsoft Edge (Chromium) in the drop-down box.

When I first read this, I did not believe you because I have been bashing my head against this problem since the beginning of the pandemic. I have just tried it out myself, and you are correct. I cannot thank you enough for writing this comment.

In case it helps anyone else, I did my experiments using the [debian "official" `chromium` package (`v114.0.5735.198-1~deb12u1`)](https://packages.debian.org/bookworm/chromium) on a debian bookworm box (yes, I do realise that this is _not_ ungoogled-chromium, but I wanted to start my testing with an environment that is easy to reproduce.)

Using the "default" user agent string[^1], clicking on the "Present now" button results in a `Your browser can't share your screen` error message.

If I then go into `Developer tools` / `Network conditions`, change my user agent (string) to `Microsoft Edge (Chromium) — Windows`[^2], and then start a new meeting _in the current tab_, clicking on the "Present now" button instead results in 3 options being presented: `Your entire screen`, `A window` or `A tab (Best for video and animation)`. I have successfully tested the last option - the audio from the shared tab **_is_** sent to the meeting clients.

I can now share video _with_ audio to my foreign language students using chromium. We are all **_very_** grateful to you for this.

[^1]:Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36
[^2]:Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko)

**Comment by whitequark** [comment]
It's beyond absolutely ridiculous, isn't it?

**Comment by valpackett** [comment]
Reminds me of how the Twitch player would just not work with a FreeBSD user agent and changing to Linux/Windows would help. They did fix it though eventually.

**Comment by tedparagon** [comment]
FWIW Brave suffers from the same issue but addresses it with an extension to opt-in to supporting Hangouts (and screen sharing).

<img width="1091" alt="image" src="https://github.com/ungoogled-software/ungoogled-chromium/assets/114427170/b33c13ac-83f1-441a-9a7c-a8954fd2eb7a">

**Comment by jaimet** [comment]
As some of the comments on this issue suggest using firefox[^1][^2][^3], I thought I'd try it out. Apologies to anyone who considers this to be "noise" - it's just that this bug report now seems to be the Internet's central repository for information about Google Meet screen sharing/"Present now" on non-chrome browsers.

**tl;dr**: You cannot currently share video **_with audio_** on meet.google.com using firefox, and I hope that this comment saves other people from wasting a day trying to get it to work.

Still reading? I installed firefox nightly v117 and then tried screen sharing/"Present now" using the default user agent string[^4]. This only offered the choice of sharing "Your entire screen" or "A window" i.e. there was no option to share video **_with audio_**.

I then modified firefox's user agent string to be every one of the 47 different user agent strings in _chromium_'s drop-down list of custom user agent strings[^5]. While changing the user agent string did result in different behaviours[^6], it did _not_ result in anything better than using firefox's _default_ user agent string.


[^1]:  ["I can use Google Meet screen sharing on Chrome and Firefox."](https://github.com/ungoogled-software/ungoogled-chromium/issues/601#issuecomment-447129671)
[^2]: ["Google meet works in firefox"](https://github.com/ungoogled-software/ungoogled-chromium/issues/601#issuecomment-785617699)
[^3]: ["Just use firefox already, out of the box work as it intended"](https://github.com/ungoogled-software/ungoogled-chromium/issues/601#issuecomment-1261975146)
[^4]: firefox nightly v117's default user agent string is `Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/117.0`
[^5]: The list of the 47 different user agent strings is in [this typescript file](https://raw.githubusercontent.com/ChromeDevTools/devtools-frontend/main/front_end/panels/network/NetworkConfigView.ts).
[^6]: Some user agent strings result in just a blank black screen when you try to start a meeting.

**Comment by Hannsek** [comment]
I bumped to this issue as well on mac. However, I have a problem with only sharing entire screen. It prompts me to grant access to chromium in system settings. Even though the permission is granted it doesn't work.
<img width="463" alt="image" src="https://github.com/ungoogled-software/ungoogled-chromium/assets/57008159/79083e64-ecf8-4232-bea8-3327dce210fb">
<img width="461" alt="image" src="https://github.com/ungoogled-software/ungoogled-chromium/assets/57008159/bfdb3674-b3c0-4e06-826e-175fbf3fe8b3">

**Comment by whitequark** [comment]
> You can make screen sharing work in Google Meet on any Chromium if you go to DevTools and change your User-Agent string to Microsoft Edge (Chromium) in the drop-down box.

Looks like they fixed this upstream as I no longer have to change User-Agent.

**Comment by Hannsek** [comment]
I missed that. Changing user-agent works for me. Thanks

**Comment by Hannsek** [comment]
> You can make screen sharing work in Google Meet on any Chromium if you go to DevTools and change your User-Agent string to Microsoft Edge (Chromium) in the drop-down box.


Ok, updating to the newest version killed this workaround and now I have the error message: 
_You can't share your screen. If you want to share a screen, you need to grant permissions_

**Comment by pascal-giguere** [comment]
FWIW, I personally stopped attempting to use ungoogled-chromium for Google Meet, and instead installed Google Meet as a PWA from Google Chrome, so I have a dedicated Google Meet icon in my Mac's dock now. Because the PWA was installed from Chrome, not Chromium, it uses Chrome under-the-hood, so screen sharing works., and I still use Chromium for everything but for Meet.

Also, I would invite ungoogled-chromium maintainers to re-consider their position regarding supporting Google Meet. I don't necessarily agree that it goes against the goals of the project ("removing Google web service dependency"):

1. Google Meet isn't a "web service dependency". It's a web app which the browser doesn't depend on, which apparently requires some workaround to work (e.g. spoofing a user agent string). That's different from having the browser itself depend on some Google service. Let's say for some reason the YouTube website stopped working on Chromium, would maintainers also not attempt to fix that on the basis that "YouTube is a Google web service dependency"? We're really stretching the definition of a "dependency" here.
2. If we're being pragmatic, the lack of proper support for a service like Google Meet, which many require for their jobs, may actually slow down the adoption of browsers like ungoogled-chromium, and ultimately make people more reliant on Google services. See my first paragraph, I didn't give up on trying to use Google Meet, I just ended up giving up using it _in ungoogled-chromium_, instead running it in Chrome. So in the end, I'm actually _more_ dependent on Google services than if ungoogled-chromium supported Meet.

Just like @pdfrod, I'd be willing to help make that support happen, but only if maintainers were on board.

**Comment by PF4Public** [comment]
> I'd be willing to help make that support happen

Did you have any progress? Perhaps this might help: https://github.com/PF4Public/gentoo-overlay/issues/314
